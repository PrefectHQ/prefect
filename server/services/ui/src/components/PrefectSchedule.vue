<script>
import { intervalToEnglish } from '@/utils/dateTime'
import cronstrue from 'cronstrue'

// TODO add button that opens a sidebar to show multiple schedules (#307)

export default {
  props: {
    schedule: {
      required: true,
      type: Object,
      validator: schedule => schedule.type
    }
  },
  computed: {
    displaySchedule() {
      if (this.schedule.type === 'CronSchedule') {
        // If there's a start date and a timezone, add a space and show it
        const timeZone =
          this.schedule.start_date && this.schedule.start_date.tz
            ? ` ${this.schedule.start_date.tz}`
            : ''
        return `${cronstrue.toString(this.schedule.cron)}${timeZone}`
      } else if (this.schedule.type === 'IntervalSchedule') {
        const microsecondsString = this.schedule.interval
        const numberOfMilliseconds = Number(microsecondsString) * 0.001

        return intervalToEnglish(numberOfMilliseconds)
      }
      return null
    }
  }
}
</script>

<template>
  <span
    v-if="
      schedule.type === 'CronSchedule' || schedule.type === 'IntervalSchedule'
    "
  >
    {{ displaySchedule }}
  </span>
  <span v-else>
    <v-btn disabled small>{{
      schedule.type === 'UnionSchedule' ? 'union schedule' : 'custom schedule'
    }}</v-btn>
  </span>
</template>
