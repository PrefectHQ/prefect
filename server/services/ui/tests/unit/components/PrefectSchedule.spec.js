import { shallowMount } from '@vue/test-utils'
import PrefectSchedule from '@/components/PrefectSchedule'

describe('PrefectSchedule', () => {
  describe('Snapshots', () => {
    it('should properly display a cron schedule with no timezone', () => {
      const wrapper = shallowMount(PrefectSchedule, {
        propsData: {
          schedule: {
            type: 'CronSchedule',
            cron: '0 12 * * *'
          }
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    it('should properly display a cron schedule with a timezone', () => {
      const wrapper = shallowMount(PrefectSchedule, {
        propsData: {
          schedule: {
            type: 'CronSchedule',
            cron: '0 12 * * *',
            start_date: {
              tz: 'EST'
            }
          }
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    it('should properly display an interval schedule', () => {
      const wrapper = shallowMount(PrefectSchedule, {
        propsData: {
          schedule: {
            type: 'IntervalSchedule',
            interval: '60000000'
          }
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    it('should properly display an Union schedule', () => {
      const wrapper = shallowMount(PrefectSchedule, {
        propsData: {
          schedule: {
            type: 'UnionSchedule'
          }
        },
        stubs: ['VBtn']
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    it('should properly display an custom schedule', () => {
      const wrapper = shallowMount(PrefectSchedule, {
        propsData: {
          schedule: {
            type: 'Schedule'
          }
        },
        stubs: ['VBtn']
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
