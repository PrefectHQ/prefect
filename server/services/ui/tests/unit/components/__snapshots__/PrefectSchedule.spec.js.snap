// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrefectSchedule Snapshots should properly display a cron schedule with a timezone 1`] = `
<span>
  
  At 12:00 PM EST

</span>
`;

exports[`PrefectSchedule Snapshots should properly display a cron schedule with no timezone 1`] = `
<span>
  
  At 12:00 PM

</span>
`;

exports[`PrefectSchedule Snapshots should properly display an Union schedule 1`] = `
<span>
  <vbtn-stub
    disabled=""
    small=""
  >
    union schedule
  </vbtn-stub>
</span>
`;

exports[`PrefectSchedule Snapshots should properly display an custom schedule 1`] = `
<span>
  <vbtn-stub
    disabled=""
    small=""
  >
    custom schedule
  </vbtn-stub>
</span>
`;

exports[`PrefectSchedule Snapshots should properly display an interval schedule 1`] = `
<span>
  
  minute

</span>
`;
